payments/schema/mutations.py

```diff
@@ -479,6 +485,77 @@ def mutate_and_get_payload(cls, root, info, **input):
         return DeleteOrderLineMutation()
 
 
+class OrderApprovalInput(graphene.InputObjectType):
+    order_id = graphene.ID(required=True)
+    email = graphene.String(required=True)
+
+
+class ApproveOrderMutation(graphene.ClientIDMutation):
+    class Input:
+        orders = graphene.List(OrderApprovalInput, required=True)
+        due_date = graphene.Date(description="Defaults to the Order due date")
+
+    failed_orders = graphene.List(FailedOrderType, required=True)
+
+    @classmethod
+    @change_permission_required(
+        Order,
+        BerthLease,
+        WinterStorageLease,
+        BerthApplication,
+        WinterStorageApplication,
+    )
+    @transaction.atomic
+    def mutate_and_get_payload(cls, root, info, **input):
+        from ..notifications import NotificationType
+
+        failed_orders = []
```

Comment:
Question(s): since the function is marked as `@transaction.atomic` all orders fail or all orders pass? can there be a case that 3 out of 5 orders will pass with data committed succesfully to db? if not, then I returning the `failed_orders` is useless.

If we would like to keep the `failed_orders` should we then adjust the transaction boundaries so that each order is processed in own separate transaction?

Reply:
very good point, we should wrap the try/except below in the atomic transaction, not whole resolver

Reply:
:big-hm: good catch. It makes sense to move the transaction to the `try/except` level.

However, after running a test in which one of them fails (raising `DoesNotExist`) and the other actually passes, the result is as expected. I.e. one `failedOrder` and one on the mailbox.

Reply:
yes, `failedOrder` will be populated, and yes mailbox will get the email, but the tables for orders, leases and applications will not get the changes for the successful ones.

Reply:
The test I added checks for
```python
assert order.due_date == due_date
assert order.lease.status == LeaseStatus.OFFERED
assert order.lease.application.status == ApplicationStatus.OFFER_SENT
```
and all of them passed 🤔 but I already moved the atomic inside, since it makes sense that it goes there.

Reply:
and yes, even the python objects in memory will be with the changes, but if you refresh them from the DB, you will see that they did not get commited 😅

Reply:
ah, no, you had tests that either had all failed or all succeed, that's why

Reply:
No no, it checks that one specific order is updated and one specific order fails. This is the full code test (without the setup)
```python
order = Order.objects.get(id=order.id)
assert len(executed["data"]["approveOrders"]["failedOrders"]) == 1
assert executed["data"]["approveOrders"]["failedOrders"][0] == {
    "id": failure_order_id,
    "error": "Order matching query does not exist.",
}
assert order.due_date == due_date
assert order.lease.status == LeaseStatus.OFFERED
assert order.lease.application.status == ApplicationStatus.OFFER_SENT
assert len(mail.outbox) == 1
assert (
    mail.outbox[0].subject
    == f"test order approved subject, event: {order.order_number}!"
)
assert mail.outbox[0].body == f"{ order.order_number } { payment_url }"
assert mail.outbox[0].to == [order.lease.application.email]
assert mail.outbox[0].alternatives == [
    (f"<b>{ order.order_number } { payment_url }</b>", "text/html")
]
```

Reply:
hm, don't see it here. do you mean you have it locally? and it works before you move the atomic to try/except level?

Reply:
Yup, I added the test locally only (I'm pushing some changes now), and it worked even _before_ moving the atomic.

Reply:
🤯 🙊

Reply:
Pushed the changes including the test